afl++ LTO variant

vanhauser-thc opened the pull request:
This should be the best generic afl++ variant, it has LTO, cmplog and a minimal MOpt variant.

These three should be default afl++ variants that run with all other fuzzers:
```
aflplusplus
aflplusplus_mopt
aflplusplus_lto_cmplog_mop2
```
all others are (so far) temporary fuzzers for individual batches.

The LTO variant requires a current gcc and the dev version of llvm 11 so the builder Dockerfile is a bit special (but I got it to work!).
However as LTO instrumentation is still bleeding edge it would be interesting if it even compiles all fuzzbench targets or if some are failing.
So if in the week you can could run a simple fuzz run (e.g. 24h, 1 run (not 20) of all fuzzbench target with this one this would be great.

vanhauser-thc commented:
I added a libfuzzer variant for valueprofiles should might give better results to libfuzzer.
why are thelibfuzzer people not adding variations? eg. the dataflow feature looks very promising.

vanhauser-thc commented:
@googlebot I consent.

vanhauser-thc commented:
one build error is a a different issue:
```
Error response from daemon: manifest for gcr.io/fuzzbench/builders/aflplusplus_lto_cmplog_mopt2/bloaty_fuzz_target-intermediate:latest not found: manifest unknown: Failed to fetch "latest" from request "/v2/fuzzbench/builders/aflplusplus_lto_cmplog_mopt2/bloaty_fuzz_target-intermediate/manifests/latest".
make: *** [.pull-aflplusplus_lto_cmplog_mopt2-bloaty_fuzz_target-oss-fuzz-builder-intermediate] Error 1
```

the compiling of the targets is then stopping at freetype. I will check if I can get it to work.

mboehme commented:
Probably related to https://github.com/google/fuzzbench/pull/291#pullrequestreview-404523730 in #291.

vanhauser-thc commented:
I dont understand the failing of the configure of freetype2.

I build my own fuzzbench build image (from builder.Dockerfile) and then used the git clone, checkout etc. commands, and for me it works fine:
```
        ./configure  '--with-harfbuzz=no' '--with-bzip2=no' '--with-png=no'
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for gcc... afl-clang-lto
checking whether the C compiler works... yes
```
and also "make" runs through without problems.

vanhauser-thc commented:
other instances also fail with
```
Error response from daemon: manifest for gcr.io/fuzzbench/builders/aflplusplus_instrim_laf_mopt2_17/bloaty_fuzz_target-intermediate:latest not found: manifest unknown: Failed to fetch "latest" from request "/v2/fuzzbench/builders/aflplusplus_instrim_laf_mopt2_17/bloaty_fuzz_target-intermediate/manifests/latest".
make: *** [.pull-aflplusplus_instrim_laf_mopt2_17-bloaty_fuzz_target-oss-fuzz-builder-intermediate] Error 1
docker/build.mk:303: recipe for target '.pull-aflplusplus_instrim_laf_mopt2_17-bloaty_fuzz_target-oss-fuzz-builder-intermediate' failed
```
what is going wrong there?

jonathanmetzman commented:
I think Abhishek fixed this issue in #291, I don't think that commit is in your branch. Can you pull please?

jonathanmetzman commented:
> The LTO variant requires a current gcc and the dev version of llvm 11 so the builder Dockerfile is a bit special (but I got it to work!).

Nice.

> However as LTO instrumentation is still bleeding edge it would be interesting if it even compiles all fuzzbench targets or if some are failing.
> So if in the week you can could run a simple fuzz run (e.g. 24h, 1 run (not 20) of all fuzzbench target with this one this would be great.

CI builds all targets. Assuming it doesn't fail spuriously this should catch build failures (though it will only catch them one at a time).

jonathanmetzman commented:
If there are builds we don't expect to work on every benchmark can we remove them from CI please?
I don't want to waste cycles and make the builds (more) red all the time.

vanhauser-thc commented:
> If there are builds we don't expect to work on every benchmark can we remove them from CI please?
> I don't want to waste cycles and make the builds (more) red all the time.
> (will document this somewhere since it came up with fastcgs).

all should build.

with LTO there is no experience so far. you said however they all built so all good to go :)

vanhauser-thc commented:
can this be merged or are there showstoppers?

jonathanmetzman commented:
> > If there are builds we don't expect to work on every benchmark can we remove them from CI please?
> > I don't want to waste cycles and make the builds (more) red all the time.
> > (will document this somewhere since it came up with fastcgs).
> 
> all should build.
> 
> with LTO there is no experience so far. you said however they all built so all good to go :)

Sorry, I didn't mean to imply they all built. I think there's clang failures for them so I assume they wont build. So let's remove from CI before landing.

jonathanmetzman commented:
> > > If there are builds we don't expect to work on every benchmark can we remove them from CI please?
> > > I don't want to waste cycles and make the builds (more) red all the time.
> > > (will document this somewhere since it came up with fastcgs).
> > 
> > 
> > all should build.
> > with LTO there is no experience so far. you said however they all built so all good to go :)
> 
> Sorry, I didn't mean to imply they all built. I think there's clang failures for them so I assume they wont build. So let's remove from CI before landing.

For example, this one: https://github.com/google/fuzzbench/pull/295/checks?check_run_id=644030980

jonathanmetzman commented:
> I added a libfuzzer variant for valueprofiles which might give better results to libfuzzer.
> why are thelibfuzzer people not adding variations? eg. the dataflow feature looks very promising.

We actually tested this pretty early on:
http://commondatastorage.googleapis.com/fuzzbench-reports/sample-and-2020-03-06-libfuzzer-only-libfuzzer/index.html

But we haven't added because it made libfuzzer worse according to fuzzbench :-)
Can you remove that addition for now?
I think the main libfuzzer devs will plan on doing more testing soon and value_profile will probably be tested.
Dataflow hasn't been tested yet because it is pretty picky about building and one of the main people working on it is out of office for a bit, but I'm sure it will happen at some point.

jonathanmetzman commented on fuzzers/aflplusplus_lto_cmplog_mopt2/Dockerfile:
```diff
@@ -0,0 +1 @@
+builder.Dockerfile
```
Sorry, I'm not sure what's happening here.

vanhauser-thc replied:
ah that was a symlink so I can easily build the container during trying to get llvm 11 and gcc 9 running in the container. just removed it and removed also libfuzzer_valueprofile

vanhauser-thc commented:
> > > If there are builds we don't expect to work on every benchmark can we remove them from CI please?
> > > I don't want to waste cycles and make the builds (more) red all the time.
> > > (will document this somewhere since it came up with fastcgs).
> > 
> > 
> > all should build.
> > with LTO there is no experience so far. you said however they all built so all good to go :)
> 
> Sorry, I didn't mean to imply they all built. I think there's clang failures for them so I assume they wont build. So let's remove from CI before landing.

How do I remove those tests that are not compiling?  so the build tests go though?

the laf-intel not working for some test cases were a bug I just fixed.

jonathanmetzman commented:
> > > > If there are builds we don't expect to work on every benchmark can we remove them from CI please?
> > > > I don't want to waste cycles and make the builds (more) red all the time.
> > > > (will document this somewhere since it came up with fastcgs).
> > > 
> > > 
> > > all should build.
> > > with LTO there is no experience so far. you said however they all built so all good to go :)
> > 
> > 
> > Sorry, I didn't mean to imply they all built. I think there's clang failures for them so I assume they wont build. So let's remove from CI before landing.
> 
> How do I remove those tests that are not compiling? so the build tests go though?
> 
> the laf-intel not working for some test cases were a bug I just fixed.

Don't worry about those. Our CI system is primitive now and tries to build everything causing a lot of spurious failures. The best we can do for now (until #278 lands) is ignore stuff that fails if you haven't touched it.

vanhauser-thc commented:
I have removed the LTO variant and added a few others. The last from he last commit did pass all tests so these variants will too.

Please merge this ones run through and start a test these:
```
          - aflplusplus_ngram3_cmplog_mopt2
          - aflplusplus_ngram3_cmplog_mopt2_17
          - aflplusplus_instrim_mopt
          - aflplusplus_instrim_mopt2
          - aflplusplus_instrim_cmplog_mopt
          - aflplusplus_instrim_cmplog_mopt2
          - aflplusplus_instrim_laf_mopt2
          - aflplusplus_instrim_laf_mopt2_17
```

For the LTO - I will do an extra PR after the merge. Is it possible to run that future PR even if some targets do not compile? I want to see which targets build, and for those that built how good the results are.

vanhauser-thc commented:
@jonathanmetzman all afl++ variants successfuly built. the one fail is aflfast - havent touched that obviously :)

vanhauser-thc commented:
@jonathanmetzman can you merge this and run a full benchmark test over the weekend? that would be great :)

jonathanmetzman commented:
I don't know that I can run the experiment this weekend but early/mid-next week should work.

vanhauser-thc commented:
OK, thanks! :)
